perses/samplers/samplers.py

```diff
@@ -1374,19 +1374,43 @@ def __init__(self, sampler, logZ=None, log_target_probabilities=None, update_met
         update_method : str, optional, default='default'
             SAMS update algorithm
         storage : NetCDFStorageView, optional, default=None
+        second_state_start : int, optional, default None
+            At what iteration number to switch to the optimal gain decay
 
         """
         # Keep copies of initializing arguments.
         # TODO: Make deep copies?
         self.sampler = sampler
+        self.chemical_states = self.sampler.proposal_engine.chemical_state_list
+
+        #Select a reference state that will always be subtracted (ensure that dict ordering does not change)
+        self._reference_state = self.chemical_states[0]
+
+        #initialize the logZ dictionary with zeroes for each chemical state
+        self.logZ = {chemical_state : 0.0 for chemical_state in self.chemical_states}
+
+        #Initialize log target probabilities with log(1/n_states)
+        self.log_target_probabilities = {chemical_state : np.log(len(self.chemical_states)) for chemical_state in self.chemical_states}
+
+        #If initial weights are specified, override any weight with what is provided
+        #However, if the chemical state is not in the reachable chemical state list,throw an exception
         if logZ is not None:
-            self.logZ = logZ
-        else:
-            self.logZ = dict()
+            for (chemical_state, logZ_value) in logZ:
+                if chemical_state not in self.chemical_states:
+                    raise ValueError("Provided a logZ initial value for an un-proposable chemical state")
+                self.logZ[chemical_state] = logZ_value
+
         if log_target_probabilities is not None:
-            self.log_target_probabilities = log_target_probabilities
-        else:
-            self.log_target_probabilities = dict()
+            for (chemical_state, log_target_probability) in log_target_probabilities:
+                if chemical_state not in self.chemical_states:
+                    raise ValueError("Provided a log target probability for an un-proposable chemical state.")
+                self.log_target_probabilities[chemical_state] = log_target_probability
+
+            #normalize target probabilities
+            #this is likely not necessary, but it is copying the algorithm in Ref 1
+            log_sum_target_probabilities = np.log(np.sum(np.exp(list(self.log_target_probabilities.values()))))
```

Comment:
Don't use this idiom! It is not numerically stable.

Use [`logsumexp`](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.misc.logsumexp.html) instead.